harvard-lts/hdc3a-dlq-handler-service
# DLQ Listeners with Resubmitting Publishers

**DLQ Listeners with Resubmitting Publishers**
* * *
_Note: [PR #1](https://github.com/harvard-lts/hdc3a-dlq-handler-service/pull/1) should be reviewed and merged first._

**GitHub Issue**: 
**([link](https://app.zenhub.com/workspaces/hdc-62163f60fb08c60011fa4795/issues/harvard-lts/hdc/184))**

**Other Relevant Links**: 
**[PR #1](https://github.com/harvard-lts/hdc3a-dlq-handler-service/pull/1)**
**[HDC-173 (Messages expiration)](https://app.zenhub.com/workspaces/hdc-62163f60fb08c60011fa4795/issues/harvard-lts/hdc/173)**

# What does this Pull Request do?
Adds DLQ Listeners for process and transfer MQs with resubmit logics through new Resubmitting Publishers.

# How should this be tested?

A description of what steps someone could take to:

* Make sure test .env variables are updated based on .env.example file
* Rebuild test environment: docker-compose -f docker-compose-test.yml build --no-cache
* Run the local test environment: `docker-compose -f docker-compose-test.yml up`
* Execute tests: `docker exec -it test_runner pytest test`

For a functional test:

* Make sure .env variables are updated based on .env.example file (_Set a small time value for MESSAGE_EXPIRATION_MS, e.g 5000 ms (5 seconds)_)
* Run the local dev environment: `docker-compose -f docker-compose-dev.yml up`
* Create a queue called "test_queue" on any of the available MQs (Transfer or Process) using the ActiveMQ web console. 
* Send a test message to the queue with the following body:
`{"admin_metadata":{"original_queue":"test_queue","retry_count":0},"test":"test"}`

You can use the ActiveMQ web console to send the message:

![activemq_send_message](https://user-images.githubusercontent.com/17010447/169472880-6574974f-6c37-435d-8522-0502682d1a0a.png)

It is important to set a value in "Time to live" so that the message expires after that time and is sent to the DLQ.

* Check that the message has been resubmitted:

Since the message sent to test_queue is not being consumed, the message will always expire even if we resubmit the message several times.

The number of times the message is resubmitted to the queue must correspond to the maximum retries set in .env.

Once the test message has been sent and having waited for it to expire for each retry, for a maximum of 3 retries, the status of the queues should look like this:

![activemq_queues_status](https://user-images.githubusercontent.com/17010447/169474272-b65bb77c-349c-42d3-80c8-2ee1bd7c56d7.png)

# Test coverage
Yes/No: Are changes in this pull-request covered by:
- unit tests Yes
- integration tests Yes

# Interested parties
@jessjass @ives1227 @adaybujeda 

Keyword: resolved
